File: integration_tests/dart_client_tests_real/test/enums_test.dart
import 'package:test/test.dart';
import 'package:graphlink_client_integration_tests/generated/client/graph_link_client.dart';
import 'package:graphlink_client_integration_tests/generated/enums/user_status.dart';
import 'package:graphlink_client_integration_tests/generated/enums/priority.dart';
import 'real_server_adapter.dart';

void main() {
  late GraphLinkClient client;
  setUp(() => client = newClient());

  group('enum deserialization', () {
    test('ACTIVE deserializes to UserStatus.active', () async {
      final res = await client.queries.getUser(id: 'user-1');
      expect(res.getUser.status, equals(UserStatus.active));
    });

    test('INACTIVE deserializes to UserStatus.inactive', () async {
      final res = await client.queries.getUser(id: 'user-2');
      expect(res.getUser.status, equals(UserStatus.inactive));
    });

    test('HIGH deserializes to Priority.high', () async {
      final res = await client.queries.getUser(id: 'user-1');
      expect(res.getUser.priority, equals(Priority.high));
    });

    test('nullable enum is null when server returns null', () async {
      final res = await client.queries.getUser(id: 'user-2');
      expect(res.getUser.priority, isNull);
    });
  });

  group('enum serialization in query variables', () {
    test('UserStatus.active filters correctly — returns only active users',
        () async {
      final res = await client.queries
          .listUsersByStatus(status: UserStatus.active);
      expect(res.listUsersByStatus, isNotEmpty);
      expect(res.listUsersByStatus.every((u) => u.status == UserStatus.active),
          isTrue);
    });

    test('UserStatus.inactive filters correctly — returns only inactive users',
        () async {
      final res = await client.queries
          .listUsersByStatus(status: UserStatus.inactive);
      expect(res.listUsersByStatus, isNotEmpty);
      expect(
          res.listUsersByStatus.every((u) => u.status == UserStatus.inactive),
          isTrue);
    });

    test('UserStatus.SUSPENDED filters correctly — returns empty list',
        () async {
      final res = await client.queries
          .listUsersByStatus(status: UserStatus.suspended);
      expect(res.listUsersByStatus, isEmpty);
    });
  });

  group('enum in list response', () {
    test('list of users contains correctly deserialized statuses', () async {
      final res = await client.queries.listUsers();
      final statuses = res.listUsers.map((u) => u.status).toList();
      expect(statuses, containsAll([UserStatus.active, UserStatus.inactive]));
    });
  });
}

File: integration_tests/typescript_client_tests_real/test/enums.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { GraphLinkClient } from '../lib/generated/client/graph-link-client.js';
import { UserStatus } from '../lib/generated/enums/user-status.js';
import { Priority } from '../lib/generated/enums/priority.js';
import { newClient } from './real-server-adapter.ts';

let client: GraphLinkClient;
beforeEach(() => { client = newClient(); });

describe('enum deserialization', () => {
  it('ACTIVE deserializes to UserStatus.Active', async () => {
    const res = await client.queries.getUser({ id: 'user-1' });
    expect(res.getUser.status).toBe(UserStatus.Active);
  });

  it('INACTIVE deserializes to UserStatus.Inactive', async () => {
    const res = await client.queries.getUser({ id: 'user-2' });
    expect(res.getUser.status).toBe(UserStatus.Inactive);
  });

  it('HIGH deserializes to Priority.High', async () => {
    const res = await client.queries.getUser({ id: 'user-1' });
    expect(res.getUser.priority).toBe(Priority.High);
  });

  it('nullable enum is null when server returns null', async () => {
    const res = await client.queries.getUser({ id: 'user-2' });
    expect(res.getUser.priority).toBeNull();
  });
});

describe('enum in list response', () => {
  it('listUsers contains correctly deserialized statuses', async () => {
    const res = await client.queries.listUsers();
    const alice = res.listUsers.find(u => u.id === 'user-1');
    const bob = res.listUsers.find(u => u.id === 'user-2');
    expect(alice?.status).toBe(UserStatus.Active);
    expect(bob?.status).toBe(UserStatus.Inactive);
  });
});

describe('enum filtering via listUsersByStatus', () => {
  it('ACTIVE returns only Alice', async () => {
    const res = await client.queries.listUsersByStatus({ status: UserStatus.Active });
    expect(res.listUsersByStatus).toHaveLength(1);
    expect(res.listUsersByStatus[0].id).toBe('user-1');
  });

  it('INACTIVE returns only Bob', async () => {
    const res = await client.queries.listUsersByStatus({ status: UserStatus.Inactive });
    expect(res.listUsersByStatus).toHaveLength(1);
    expect(res.listUsersByStatus[0].id).toBe('user-2');
  });

  it('SUSPENDED returns empty list', async () => {
    const res = await client.queries.listUsersByStatus({ status: UserStatus.Suspended });
    expect(res.listUsersByStatus).toHaveLength(0);
  });
});
